File: taurus-protect-sdk-python/taurus_protect/models/request.py
"""Request models for Taurus-PROTECT SDK."""

from __future__ import annotations

import json
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from taurus_protect.models.currency import Currency


class RequestStatus(str, Enum):
    """
    Status of a transaction request.

    Request statuses track the progress of a request from creation through
    approval, signing, broadcasting, and confirmation on the blockchain.

    Common status transitions:
    - CREATED -> PENDING -> APPROVED -> BROADCASTED -> CONFIRMED
    - PENDING -> REJECTED (if rejected by an approver)
    - BROADCASTED -> PERMANENT_FAILURE (if transaction fails)
    """

    # Secondary approval received
    APPROVED_2 = "APPROVED_2"
    # Request has been approved and is ready for signing
    APPROVED = "APPROVED"
    # Request is in the process of being approved
    APPROVING = "APPROVING"
    # Auto prepared secondary status
    AUTO_PREPARED_2 = "AUTO_PREPARED_2"
    # Auto prepared status
    AUTO_PREPARED = "AUTO_PREPARED"
    # Broadcasting secondary status
    BROADCASTING_2 = "BROADCASTING_2"
    # Broadcasting status
    BROADCASTING = "BROADCASTING"
    # Broadcasted to the blockchain
    BROADCASTED = "BROADCASTED"
    # Bundle has been approved
    BUNDLE_APPROVED = "BUNDLE_APPROVED"
    # Bundle is being broadcast
    BUNDLE_BROADCASTING = "BUNDLE_BROADCASTING"
    # Bundle is ready
    BUNDLE_READY = "BUNDLE_READY"
    # Request was canceled
    CANCELED = "CANCELED"
    # Transaction confirmed on blockchain
    CONFIRMED = "CONFIRMED"
    # Request has been created
    CREATED = "CREATED"
    # Diem burn MBS approved
    DIEM_BURN_MBS_APPROVED = "DIEM_BURN_MBS_APPROVED"
    # Diem burn MBS pending
    DIEM_BURN_MBS_PENDING = "DIEM_BURN_MBS_PENDING"
    # Diem mint MBS approved
    DIEM_MINT_MBS_APPROVED = "DIEM_MINT_MBS_APPROVED"
    # Diem mint MBS completed
    DIEM_MINT_MBS_COMPLETED = "DIEM_MINT_MBS_COMPLETED"
    # Diem mint MBS pending
    DIEM_MINT_MBS_PENDING = "DIEM_MINT_MBS_PENDING"
    # Request has expired
    EXPIRED = "EXPIRED"
    # Fast approved secondary status
    FAST_APPROVED_2 = "FAST_APPROVED_2"
    # HSM signing failed (secondary)
    HSM_FAILED_2 = "HSM_FAILED_2"
    # HSM signing failed
    HSM_FAILED = "HSM_FAILED"
    # HSM ready for signing (secondary)
    HSM_READY_2 = "HSM_READY_2"
    # HSM ready for signing
    HSM_READY = "HSM_READY"
    # HSM signed (secondary)
    HSM_SIGNED_2 = "HSM_SIGNED_2"
    # HSM has signed the transaction
    HSM_SIGNED = "HSM_SIGNED"
    # Manual broadcast required
    MANUAL_BROADCAST = "MANUAL_BROADCAST"
    # Transaction has been mined
    MINED = "MINED"
    # Transaction partially confirmed
    PARTIALLY_CONFIRMED = "PARTIALLY_CONFIRMED"
    # Request is pending approval
    PENDING = "PENDING"
    # Transaction permanently failed
    PERMANENT_FAILURE = "PERMANENT_FAILURE"
    # Request is ready
    READY = "READY"
    # Request was rejected
    REJECTED = "REJECTED"
    # Transaction has been sent
    SENT = "SENT"
    # Signet transaction completed
    SIGNET_COMPLETED = "SIGNET_COMPLETED"
    # Signet transaction pending
    SIGNET_PENDING = "SIGNET_PENDING"
    # Unknown status
    UNKNOWN = "UNKNOWN"

    @classmethod
    def from_string(cls, value: str) -> "RequestStatus":
        """Convert string to RequestStatus, with fallback to UNKNOWN."""
        if not value:
            return cls.UNKNOWN
        try:
            return cls(value.upper())
        except ValueError:
            return cls.UNKNOWN


class RequestTrail(BaseModel):
    """Audit trail entry for a request."""

    timestamp: Optional[datetime] = Field(default=None, description="When the action occurred")
    action: Optional[str] = Field(default=None, description="Action taken")
    user_id: Optional[str] = Field(default=None, description="User who took action")
    comment: Optional[str] = Field(default=None, description="Optional comment")

    model_config = {"frozen": True}


class RequestMetadataAmount(BaseModel):
    """Amount information extracted from request metadata payload."""

    value_from: Optional[str] = None
    value_to: Optional[str] = None
    rate: Optional[str] = None
    decimals: Optional[int] = None  # Always a small integer
    currency_from: Optional[str] = None
    currency_to: Optional[str] = None

    model_config = ConfigDict(frozen=True)


class RequestMetadata(BaseModel):
    """
    Metadata for a transaction request.

    Contains only the hash and the verified payload string used for
    integrity verification. Use the ``get_*()`` convenience methods
    to extract fields from the cryptographically verified payload.

    SECURITY DESIGN:
        - ``hash``: SHA-256 of ``payload_as_string``, verified against governance rules
        - ``payload_as_string``: canonical JSON used for hash computation (VERIFIED)
        - ``payload`` (raw JSON object): intentionally omitted to prevent use of
          unverified data. An attacker could modify the raw payload while leaving
          ``payload_as_string`` unchanged — the hash would still verify, but the
          client would extract tampered data.
    """

    hash: Optional[str] = Field(default=None, description="SHA-256 hash of payload_as_string")
    payload_as_string: Optional[str] = Field(default=None, description="Verified JSON payload")

    model_config = {"frozen": True}

    def _parse_payload_entries(self) -> list:
        """Parse payload_as_string as JSON array of key-value entries."""
        if not self.payload_as_string:
            return []
        try:
            entries = json.loads(self.payload_as_string)
            return entries if isinstance(entries, list) else []
        except (json.JSONDecodeError, TypeError):
            return []

    def _get_payload_value(self, key: str) -> Any:
        """Get the value for a given key from payload entries."""
        for entry in self._parse_payload_entries():
            if isinstance(entry, dict) and entry.get("key") == key:
                return entry.get("value")
        return None

    def get_source_address(self) -> Optional[str]:
        """Extract source address from verified payload."""
        value = self._get_payload_value("source")
        if isinstance(value, dict):
            payload = value.get("payload")
            if isinstance(payload, dict):
                return payload.get("address")
        return None

    def get_destination_address(self) -> Optional[str]:
        """Extract destination address from verified payload."""
        value = self._get_payload_value("destination")
        if isinstance(value, dict):
            payload = value.get("payload")
            if isinstance(payload, dict):
                return payload.get("address")
        return None

    def get_amount(self) -> Optional[RequestMetadataAmount]:
        """Extract amount information from verified payload."""
        value = self._get_payload_value("amount")
        if not isinstance(value, dict):
            return None
        return RequestMetadataAmount(
            value_from=_json_value_to_string(value.get("valueFrom")),
            value_to=_json_value_to_string(value.get("valueTo")),
            rate=_json_value_to_string(value.get("rate")),
            decimals=int(value["decimals"]) if "decimals" in value else None,
            currency_from=str(value["currencyFrom"]) if "currencyFrom" in value else None,
            currency_to=str(value["currencyTo"]) if "currencyTo" in value else None,
        )


def _json_value_to_string(value: Any) -> Optional[str]:
    """Convert a JSON value to string, handling both string and numeric inputs.

    The API returns valueFrom, valueTo, and rate as JSON strings to support
    arbitrary-precision amounts exceeding 64-bit limits. This helper handles
    both string and numeric JSON values for backward compatibility.
    """
    if value is None:
        return None
    if isinstance(value, str):
        return value
    if isinstance(value, float):
        return f"{value:.20g}"
    return str(value)


class Attribute(BaseModel):
    """Custom key-value attribute that can be attached to various entities."""

    id: Optional[int] = Field(default=None, description="Attribute identifier")
    key: Optional[str] = Field(default=None, description="Attribute key")
    value: Optional[str] = Field(default=None, description="Attribute value")
    content_type: Optional[str] = Field(default=None, description="Content type for file attributes")
    owner: Optional[str] = Field(default=None, description="Owner identifier")
    type: Optional[str] = Field(default=None, description="Type classification")
    sub_type: Optional[str] = Field(default=None, description="Sub-type classification")
    is_file: bool = Field(default=False, description="Whether the value is a file reference")

    model_config = ConfigDict(frozen=True)


class RequestApproversGroup(BaseModel):
    """Single approval group requirement for a request."""

    external_group_id: Optional[str] = Field(default=None, description="External group identifier")
    minimum_signatures: int = Field(default=0, description="Minimum signatures required")

    model_config = ConfigDict(frozen=True)


class RequestParallelApproversGroups(BaseModel):
    """Set of approval groups that operate in parallel."""

    sequential: List[RequestApproversGroup] = Field(
        default_factory=list, description="Sequential approval groups"
    )

    model_config = ConfigDict(frozen=True)


class RequestApprovers(BaseModel):
    """Approvers configuration for a transaction request."""

    parallel: List[RequestParallelApproversGroups] = Field(
        default_factory=list, description="Parallel approver groups"
    )

    model_config = ConfigDict(frozen=True)


class SignedRequest(BaseModel):
    """A signed blockchain transaction associated with a request."""

    id: Optional[str] = None
    signed_request: Optional[str] = None
    status: Optional[RequestStatus] = None
    hash: Optional[str] = None
    block: Optional[int] = None
    details: Optional[str] = None
    creation_date: Optional[datetime] = None
    update_date: Optional[datetime] = None
    broadcast_date: Optional[datetime] = None
    confirmation_date: Optional[datetime] = None

    model_config = ConfigDict(frozen=True)


class Request(BaseModel):
    """
    Transaction request.

    Represents a pending or completed transaction request with full
    audit trail and metadata for integrity verification.
    """

    id: str = Field(description="Unique request identifier")
    tenant_id: Optional[int] = Field(default=None, description="Tenant identifier")
    currency: Optional[str] = Field(default=None, description="Currency symbol")
    envelope: Optional[str] = Field(default=None, description="Unsigned transaction")
    status: RequestStatus = Field(default=RequestStatus.PENDING, description="Request status")
    trails: List[RequestTrail] = Field(default_factory=list, description="Audit trail")
    created_at: Optional[datetime] = Field(default=None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(default=None, description="Update timestamp")
    metadata: Optional[RequestMetadata] = Field(default=None, description="Request metadata")
    rule: Optional[str] = Field(default=None, description="Applied governance rule")
    signed_requests: List[SignedRequest] = Field(default_factory=list, description="Signed transactions")
    type: Optional[str] = Field(default=None, description="Request type")
    approvers: Optional[RequestApprovers] = Field(default=None, description="Approval configuration")
    currency_info: Optional[Currency] = Field(default=None, description="Detailed currency information")
    needs_approval_from: List[str] = Field(default_factory=list, description="Pending approvers")
    request_bundle_id: Optional[str] = Field(default=None, description="Request bundle ID")
    external_request_id: Optional[str] = Field(default=None, description="External request ID")
    attributes: List[Attribute] = Field(default_factory=list, description="Custom attributes")

    model_config = {"frozen": True}


class CreateInternalTransferRequest(BaseModel):
    """Request to create an internal transfer between addresses."""

    from_address_id: str = Field(description="Source address ID")
    to_address_id: str = Field(description="Destination address ID")
    amount: str = Field(description="Amount to transfer")
    comment: Optional[str] = Field(default=None, description="Optional comment")
    external_request_id: Optional[str] = Field(default=None, description="External request ID")
    fee_level: Optional[str] = Field(default=None, description="Fee level (low, medium, high)")


class CreateExternalTransferRequest(BaseModel):
    """Request to create an external transfer to a whitelisted address."""

    from_address_id: str = Field(description="Source address ID")
    to_whitelisted_address_id: str = Field(description="Destination whitelisted address ID")
    amount: str = Field(description="Amount to transfer")
    comment: Optional[str] = Field(default=None, description="Optional comment")
    external_request_id: Optional[str] = Field(default=None, description="External request ID")
    fee_level: Optional[str] = Field(default=None, description="Fee level (low, medium, high)")
    memo: Optional[str] = Field(default=None, description="Optional memo/destination tag")


class ListRequestsOptions(BaseModel):
    """Options for listing requests."""

    limit: int = Field(default=50, ge=1, le=1000, description="Maximum items to return")
    offset: int = Field(default=0, ge=0, description="Number of items to skip")
    status: Optional[RequestStatus] = Field(default=None, description="Filter by status")
    currency: Optional[str] = Field(default=None, description="Filter by currency")
    wallet_id: Optional[str] = Field(default=None, description="Filter by wallet ID")

File: taurus-protect-sdk-typescript/src/models/request.ts
/**
 * Status of a transaction request.
 *
 * Request statuses track the progress of a request from creation through
 * approval, signing, broadcasting, and confirmation on the blockchain.
 *
 * Common status transitions:
 * - CREATED -> PENDING -> APPROVED -> BROADCASTED -> CONFIRMED
 * - PENDING -> REJECTED (if rejected by an approver)
 * - BROADCASTED -> PERMANENT_FAILURE (if transaction fails)
 */
export enum RequestStatus {
  /** Request has received secondary approval */
  APPROVED_2 = "APPROVED_2",
  /** Request has been approved and is ready for signing */
  APPROVED = "APPROVED",
  /** Request is in the process of being approved */
  APPROVING = "APPROVING",
  /** Auto prepared secondary status */
  AUTO_PREPARED_2 = "AUTO_PREPARED_2",
  /** Auto prepared status */
  AUTO_PREPARED = "AUTO_PREPARED",
  /** Broadcasting secondary status */
  BROADCASTING_2 = "BROADCASTING_2",
  /** Broadcasting status */
  BROADCASTING = "BROADCASTING",
  /** Broadcasted to the blockchain */
  BROADCASTED = "BROADCASTED",
  /** Bundle has been approved */
  BUNDLE_APPROVED = "BUNDLE_APPROVED",
  /** Bundle is being broadcast */
  BUNDLE_BROADCASTING = "BUNDLE_BROADCASTING",
  /** Bundle is ready */
  BUNDLE_READY = "BUNDLE_READY",
  /** Request was canceled */
  CANCELED = "CANCELED",
  /** Transaction confirmed on blockchain */
  CONFIRMED = "CONFIRMED",
  /** Request has been created */
  CREATED = "CREATED",
  /** Diem burn MBS approved */
  DIEM_BURN_MBS_APPROVED = "DIEM_BURN_MBS_APPROVED",
  /** Diem burn MBS pending */
  DIEM_BURN_MBS_PENDING = "DIEM_BURN_MBS_PENDING",
  /** Diem mint MBS approved */
  DIEM_MINT_MBS_APPROVED = "DIEM_MINT_MBS_APPROVED",
  /** Diem mint MBS completed */
  DIEM_MINT_MBS_COMPLETED = "DIEM_MINT_MBS_COMPLETED",
  /** Diem mint MBS pending */
  DIEM_MINT_MBS_PENDING = "DIEM_MINT_MBS_PENDING",
  /** Request has expired */
  EXPIRED = "EXPIRED",
  /** Fast approved secondary status */
  FAST_APPROVED_2 = "FAST_APPROVED_2",
  /** HSM signing failed (secondary) */
  HSM_FAILED_2 = "HSM_FAILED_2",
  /** HSM signing failed */
  HSM_FAILED = "HSM_FAILED",
  /** HSM ready for signing (secondary) */
  HSM_READY_2 = "HSM_READY_2",
  /** HSM ready for signing */
  HSM_READY = "HSM_READY",
  /** HSM signed (secondary) */
  HSM_SIGNED_2 = "HSM_SIGNED_2",
  /** HSM has signed the transaction */
  HSM_SIGNED = "HSM_SIGNED",
  /** Manual broadcast required */
  MANUAL_BROADCAST = "MANUAL_BROADCAST",
  /** Transaction has been mined */
  MINED = "MINED",
  /** Transaction partially confirmed */
  PARTIALLY_CONFIRMED = "PARTIALLY_CONFIRMED",
  /** Request is pending approval */
  PENDING = "PENDING",
  /** Transaction permanently failed */
  PERMANENT_FAILURE = "PERMANENT_FAILURE",
  /** Request is ready */
  READY = "READY",
  /** Request was rejected */
  REJECTED = "REJECTED",
  /** Transaction has been sent */
  SENT = "SENT",
  /** Signet transaction completed */
  SIGNET_COMPLETED = "SIGNET_COMPLETED",
  /** Signet transaction pending */
  SIGNET_PENDING = "SIGNET_PENDING",
  /** Unknown status */
  UNKNOWN = "UNKNOWN",
}

/**
 * Request type enum.
 */
export enum RequestType {
  UNKNOWN = "UNKNOWN",
  TRANSFER = "TRANSFER",
  SIGN = "SIGN",
  CONTRACT_CALL = "CONTRACT_CALL",
  PAYMENT = "payment",
  FUNCTION_CALL = "function_call",
}

/**
 * Request metadata containing cryptographic hash information.
 *
 * The metadata is used for integrity verification:
 * - hash: SHA-256 hash of the payloadAsString
 * - payloadAsString: The canonical string representation used for hash computation
 *
 * SECURITY: payload field intentionally omitted - use payloadAsString only.
 * The raw payload object could be tampered with by an attacker while
 * payloadAsString remains unchanged (hash still verifies). By not having
 * this field, we enforce that all data extraction uses the verified source.
 */
export interface RequestMetadata {
  /** SHA-256 hash of the request payload (hex-encoded) */
  readonly hash: string;
  /** The raw payload used for hash computation */
  readonly payloadAsString: string | undefined;
  // SECURITY: payload field intentionally omitted - use payloadAsString only.
  // Use JSON.parse(payloadAsString) for secure data extraction.
}

/** A signed blockchain transaction associated with a request. */
export interface SignedRequest {
  readonly id: string;
  readonly signedRequest: string;
  readonly status: RequestStatus | string;
  readonly hash: string;
  readonly block: number;
  readonly details: string;
  readonly creationDate: Date | undefined;
  readonly updateDate: Date | undefined;
  readonly broadcastDate: Date | undefined;
  readonly confirmationDate: Date | undefined;
}

/** Amount information extracted from request metadata payload. */
export interface RequestMetadataAmount {
  readonly valueFrom: string;
  readonly valueTo: string;
  readonly rate: string;
  readonly decimals: number;
  readonly currencyFrom: string;
  readonly currencyTo: string;
}

/**
 * Represents a transaction request in Taurus-PROTECT.
 *
 * Requests are cryptographically protected:
 * - The metadata.hash field contains a SHA-256 hash of the request payload
 * - Hash verification should use constant-time comparison
 * - Approval requires ECDSA signing of the hash
 *
 * **Security Features:**
 * - All requests fetched via RequestService.get() have their hash verified
 * - Request approval uses ECDSA signatures for cryptographic verification
 */
export interface Request {
  /** Unique identifier */
  readonly id: number;
  /** Request type */
  readonly type: RequestType | string;
  /** Current status */
  readonly status: RequestStatus | string;
  /** Metadata containing hash and payload for verification */
  readonly metadata: RequestMetadata | undefined;
  /** Tenant ID */
  readonly tenantId: string | undefined;
  /** Currency/asset identifier */
  readonly currency: string | undefined;
  /** Currency information */
  readonly currencyInfo: CurrencyInfo | undefined;
  /** Request memo/reference */
  readonly memo: string | undefined;
  /** Request rule applied */
  readonly rule: string | undefined;
  /** External request ID for idempotency */
  readonly externalRequestId: string | undefined;
  /** Request bundle ID */
  readonly requestBundleId: string | undefined;
  /** Groups that need to approve this request */
  readonly needsApprovalFrom: string[];
  /** Approvers information */
  readonly approvers: Approvers | undefined;
  /** Creation timestamp */
  readonly createdAt: Date | undefined;
  /** Last update timestamp */
  readonly updatedAt: Date | undefined;
  /** Associated tags */
  readonly tags: string[];
  /** Signed blockchain transactions associated with this request */
  readonly signedRequests: SignedRequest[];
}

/**
 * Currency information.
 */
export interface CurrencyInfo {
  readonly id: string | undefined;
  readonly symbol: string | undefined;
  readonly name: string | undefined;
  readonly decimals: number | undefined;
  readonly blockchain: string | undefined;
  readonly network: string | undefined;
}

/**
 * Approvers group specifying a group and required signature count.
 */
export interface ApproversGroup {
  /** External group ID */
  readonly externalGroupId: string | undefined;
  /** Minimum signatures required from this group */
  readonly minimumSignatures: number | undefined;
}

/**
 * Sequential approvers groups that must approve in order.
 */
export interface ParallelApproversGroups {
  /** Sequential groups */
  readonly sequential: ApproversGroup[];
}

/**
 * Approvers information for a request.
 *
 * The approval structure supports parallel and sequential approval groups,
 * allowing complex multi-signature approval workflows.
 */
export interface Approvers {
  /** Parallel approval groups */
  readonly parallel: ParallelApproversGroups[];
}

/**
 * Options for listing requests.
 */
export interface ListRequestsOptions {
  /** Maximum number of requests to return */
  limit?: number;
  /** Page to request (for cursor-based pagination) */
  pageRequest?: "FIRST" | "PREVIOUS" | "NEXT" | "LAST";
  /** Current page cursor */
  currentPage?: string;
  /** Filter by request statuses */
  statuses?: RequestStatus[];
  /** Filter by request types */
  types?: string[];
  /** Filter by request IDs */
  ids?: number[];
  /** Filter by currency ID */
  currencyId?: string;
  /** Filter requests created after this date */
  fromDate?: Date;
  /** Filter requests created before this date */
  toDate?: Date;
  /** Sort order */
  sortOrder?: "ASC" | "DESC";
  /** Filter by external request IDs */
  externalRequestIds?: string[];
}

/**
 * Options for listing requests pending approval.
 */
export interface ListRequestsForApprovalOptions {
  /** Maximum number of requests to return */
  limit?: number;
  /** Page to request (for cursor-based pagination) */
  pageRequest?: "FIRST" | "PREVIOUS" | "NEXT" | "LAST";
  /** Current page cursor */
  currentPage?: string;
  /** Filter by request types */
  types?: string[];
  /** Exclude request types */
  excludeTypes?: string[];
  /** Filter by request IDs */
  ids?: number[];
  /** Filter by currency ID */
  currencyId?: string;
  /** Sort order */
  sortOrder?: "ASC" | "DESC";
  /** Filter by external request IDs */
  externalRequestIds?: string[];
}

/**
 * Options for creating an internal transfer request.
 */
export interface CreateInternalTransferOptions {
  /** Source address ID */
  fromAddressId: number;
  /** Destination address ID */
  toAddressId: number;
  /** Transfer amount as string (to preserve precision) */
  amount: string;
  /** Optional comment */
  comment?: string;
  /** Optional external request ID for idempotency */
  externalRequestId?: string;
  /** Optional gas limit */
  gasLimit?: string;
  /** Optional fee limit */
  feeLimit?: string;
}

/**
 * Options for creating an external transfer request.
 */
export interface CreateExternalTransferOptions {
  /** Source address ID */
  fromAddressId: number;
  /** Destination whitelisted address ID */
  toWhitelistedAddressId: number;
  /** Transfer amount as string (to preserve precision) */
  amount: string;
  /** Optional comment */
  comment?: string;
  /** Optional external request ID for idempotency */
  externalRequestId?: string;
  /** Optional gas limit */
  gasLimit?: string;
  /** Optional fee limit */
  feeLimit?: string;
  /** Optional destination address memo */
  destinationAddressMemo?: string;
}

/**
 * Options for creating an internal transfer from a wallet.
 */
export interface CreateInternalTransferFromWalletOptions {
  /** Source wallet ID (must be an omnibus wallet) */
  fromWalletId: number;
  /** Destination address ID */
  toAddressId: number;
  /** Transfer amount as string (to preserve precision) */
  amount: string;
  /** Optional comment */
  comment?: string;
  /** Optional external request ID for idempotency */
  externalRequestId?: string;
  /** Optional gas limit */
  gasLimit?: string;
  /** Optional fee limit */
  feeLimit?: string;
}

/**
 * Options for creating an external transfer from a wallet.
 */
export interface CreateExternalTransferFromWalletOptions {
  /** Source wallet ID (must be an omnibus wallet) */
  fromWalletId: number;
  /** Destination whitelisted address ID */
  toWhitelistedAddressId: number;
  /** Transfer amount as string (to preserve precision) */
  amount: string;
  /** Optional comment */
  comment?: string;
  /** Optional external request ID for idempotency */
  externalRequestId?: string;
  /** Optional gas limit */
  gasLimit?: string;
  /** Optional fee limit */
  feeLimit?: string;
  /** Optional destination address memo */
  destinationAddressMemo?: string;
}

/**
 * Options for creating an incoming request from an exchange.
 */
export interface CreateIncomingRequestOptions {
  /** Source exchange ID */
  fromExchangeId: number;
  /** Destination address ID */
  toAddressId: number;
  /** Transfer amount as string (to preserve precision) */
  amount: string;
  /** Optional comment */
  comment?: string;
  /** Optional external request ID for idempotency */
  externalRequestId?: string;
}
